lib/membrane_mp4/demuxer/isom/engine.ex

```diff
@@ -0,0 +1,197 @@
+
+  @typedoc """
+  A type representing a callback that is used to provide data to the demuxer.
+  The callback needs to accept the start position and size (both expressed in bytes) and 
+  needs to return a binary of that size.
+  """
+  @type provide_data_cb :: (start :: non_neg_integer(), size :: pos_integer(), state :: any() ->
+                              {data :: binary(), state :: any()})
+
+  @typedoc """
+  A type representing the `#{inspect(__MODULE__)}`.
+  """
+  @opaque t :: %__MODULE__{
+            provide_data_cb: provide_data_cb(),
+            cursor: non_neg_integer(),
+            box_positions: %{
+              Container.box_name_t() =>
+                {offset :: non_neg_integer(), header_size :: pos_integer(),
+                 content_size :: pos_integer()}
+            },
+            boxes: Container.t(),
+            tracks: %{
+              samples: [
+                %{
+                  size: pos_integer(),
+                  sample_delta: pos_integer(),
+                  track_id: pos_integer()
+                }
+              ]
+            },
+            samples_info: SamplesInfo.t(),
+            provider_state: any()
+          }
+
+  @enforce_keys [:provide_data_cb]
+  defstruct @enforce_keys ++
+              [
+                cursor: 0,
+                box_positions: %{},
+                boxes: [],
+                tracks: %{},
+                samples_info: %{},
+                provider_state: nil
+              ]
+
+  @max_header_size 16
+
+  @doc """
+  Returns new instance of the `#{inspect(__MODULE__)}`.
+  """
+  @spec new(provide_data_cb()) :: t()
+  def new(provide_data_cb) do
+    %__MODULE__{provide_data_cb: provide_data_cb}
+    |> find_box(:moov)
+    |> read_box(:moov)
+    |> find_box(:mdat)
+    |> resolve_samples()
+    |> initialize_tracks()
+  end
+
+  @doc """
+  Returns a map describing tracks found in the MP4 file.
+  """
+  @spec get_tracks_info(t()) :: %{(track_id :: pos_integer()) => SampleTable.t()}
+  def(get_tracks_info(state)) do
+    state.samples_info.sample_tables
+  end
+
+  @doc """
+  Reads the sample from given track.
+  """
+  @spec read_sample(t(), track_id :: pos_integer()) :: {:ok, Sample.t(), t()} | :end_of_stream
+  def read_sample(state, track_id) do
+    next_dts = state.tracks[track_id].next_dts
+
+    case state.tracks[track_id].samples do
+      [] ->
+        :end_of_stream
+
+      [sample | rest] ->
+        size = sample.size
+        offset = sample.sample_offset
+        {data, provider_state} = state.provide_data_cb.(offset, size, state.provider_state)
+        state = put_in(state.provider_state, provider_state)
+
+        dts = next_dts
+        pts = dts + sample.sample_composition_offset
+
+        state = put_in(state.tracks[track_id].next_dts, dts + sample.sample_delta)
+        pts_ms = (pts / state.samples_info.timescales[track_id] * 1000) |> round()
+        dts_ms = (dts / state.samples_info.timescales[track_id] * 1000) |> round()
```

Comment:
won't this introduce rounding errors in the long run?

Reply:
the error shouldn't accumulate (`next_dts` field is an integer and we only perform rounding when "casting" to `dts_ms` and `pts_ms`)